server/src/main/java/org/elasticsearch/common/time/EpochTime.java

```diff
@@ -169,10 +192,18 @@ public TemporalAccessor resolve(
                 nanos = secondsAndMillis % 1000 * 1_000_000;
```

Comment:
Taking a fresh look at this, I wonder if this line is the root of the problem. Why do we extract nanos _after_ making `secondsAndMillis` negative? If we extracted the millis portion before negation, it would always be positive, and we could always add nanos to it? I think it would be much easier to reason about if we didn't have these intermediate negatives since ultimately nanos must be positive.